src/main/java/org/opensearch/plugin/insights/core/service/grouper/MinMaxHeapQueryGrouper.java

```diff
@@ -0,0 +1,298 @@
+     */
+    @Override
+    public boolean setGroupingType(GroupingType newGroupingType) {
+        if (this.groupingType != newGroupingType) {
+            this.groupingType = newGroupingType;
+            drain();
+            return true;
+        }
+        return false;
+    }
+
+    /**
+     * Get Grouping Type
+     * @return grouping type
+     */
+    @Override
+    public GroupingType getGroupingType() {
+        return groupingType;
+    }
+
+    /**
+     * Set the maximum number of groups that should be tracked when calculating Top N groups.
+     * If the value changes, reset the state of the query grouper service by draining all internal data.
+     * @param maxGroups max number of groups
+     * @return max groups changed
+     */
+    @Override
+    public boolean setMaxGroups(int maxGroups) {
+        if (this.maxGroups != maxGroups) {
+            this.maxGroups = maxGroups;
+            drain();
+            return true;
+        }
+        return false;
+    }
+
+    /**
+     * Update Top N size
+     * @param newSize new size
+     */
+    @Override
+    public void updateTopNSize(int newSize) {
+        this.topNSize = newSize;
+    }
+
+    private void addToMinPQOverflowToMaxPQ(SearchQueryRecord searchQueryRecord, String groupId) {
+        minHeapTopQueriesStore.add(searchQueryRecord);
+        groupIdToAggSearchQueryRecord.put(groupId, new Tuple<>(searchQueryRecord, true));
+
+        while (minHeapTopQueriesStore.size() > topNSize) {
+            SearchQueryRecord recordMovedFromMinToMax = minHeapTopQueriesStore.poll();
+            maxHeapQueryStore.add(recordMovedFromMinToMax);
+            groupIdToAggSearchQueryRecord.put(recordMovedFromMinToMax.getGroupingId(), new Tuple<>(recordMovedFromMinToMax, false));
+        }
+    }
+
+    private void updateToMaxPQ(SearchQueryRecord searchQueryRecord, SearchQueryRecord aggregateSearchQueryRecord, String groupId) {
+        maxHeapQueryStore.remove(aggregateSearchQueryRecord);
+        Number measurementToAdd = searchQueryRecord.getMeasurement(metricType);
+        aggregateSearchQueryRecord.addMeasurement(metricType, measurementToAdd);
+
+        if (minHeapTopQueriesStore.size() >= topNSize) {
+            addToMinPQOverflowToMaxPQ(aggregateSearchQueryRecord, groupId);
+        } else {
+            addToMaxPQPromoteToMinPQ(aggregateSearchQueryRecord, groupId);
+        }
+    }
+
+    private void updateToMinPQ(SearchQueryRecord searchQueryRecord, SearchQueryRecord aggregateSearchQueryRecord, String groupId) {
+        minHeapTopQueriesStore.remove(aggregateSearchQueryRecord);
+        Number measurementToAdd = searchQueryRecord.getMeasurement(metricType);
+        aggregateSearchQueryRecord.addMeasurement(metricType, measurementToAdd);
+
+        if (maxHeapQueryStore.size() > 0) {
+            addToMaxPQPromoteToMinPQ(aggregateSearchQueryRecord, groupId);
+        } else {
+            addToMinPQOverflowToMaxPQ(aggregateSearchQueryRecord, groupId);
+        }
```

Comment:
Why do we need to do this `if/else` here? Can't we simply _always_ add to max/min queue and then do the promote/overflow? 
Also the "else" logic looks effectiveness to me. If the execution enters this else, that means we have already removed a record from the min queue, and max queue is also empty. So when we do `addToMinPQOverflowToMaxPQ` we are adding the previous (but updated) record back it to the min queue again and we won't overflow anything to max queue.

Reply:
The whole logic can be simplied with:
```
public SearchQueryRecord add(SearchQueryRecord searchQueryRecord) {
    if (!groupIdToAggSearchQueryRecord.containsKey(groupId)) {
        boolean maxGroupsLimitReached = checkMaxGroupsLimitReached(groupId);
        if (maxGroupsLimitReached) {
            return null;
        }
        aggregateSearchQueryRecord = searchQueryRecord;
        aggregateSearchQueryRecord.setGroupingId(groupId);
        aggregateSearchQueryRecord.setMeasurementAggregation(metricType, aggregationType);
        addToMinPQ(aggregateSearchQueryRecord, groupId);
    } else {
        aggregateSearchQueryRecord = groupIdToAggSearchQueryRecord.get(groupId).v1();
        boolean isPresentInMinPQ = groupIdToAggSearchQueryRecord.get(groupId).v2();
        if (isPresentInMinPQ) {
            minHeapTopQueriesStore.remove(aggregateSearchQueryRecord);
        } else {
            maxHeapTopQueriesStore.remove(aggregateSearchQueryRecord);
        }
        addAndPromote(searchQueryRecord, aggregateSearchQueryRecord, groupId);
        
    }
    return aggregateSearchQueryRecord;
}
private void addToMinPQ(SearchQueryRecord searchQueryRecord, String groupId) {
    minHeapTopQueriesStore.add(searchQueryRecord);
    groupIdToAggSearchQueryRecord.put(groupId, new Tuple<>(searchQueryRecord, true));
    overflow();
}
private void addAndPromote(SearchQueryRecord searchQueryRecord, SearchQueryRecord aggregateSearchQueryRecord, String groupId) {
    Number measurementToAdd = searchQueryRecord.getMeasurement(metricType);
    aggregateSearchQueryRecord.addMeasurement(metricType, measurementToAdd);
    addToMinPQ(aggregateSearchQueryRecord, groupId);
    if (maxHeapQueryStore.isEmpty()) {
        return;
    } 
    if (SearchQueryRecord.compare(maxHeapQueryStore.peek(), minHeapTopQueriesStore.peak(), metricType) > 0) {
        SearchQueryRecord recordMovedFromMaxToMin = maxHeapQueryStore.poll();
        addToMinPQ(recordMovedFromMaxToMin, recordMovedFromMaxToMin.getGroupingId());
    }
}
private void overflow() {
    if (minHeapTopQueriesStore.size() > topNSize) {
        SearchQueryRecord recordMovedFromMinToMax = minHeapTopQueriesStore.poll();
        maxHeapQueryStore.add(recordMovedFromMinToMax);
        groupIdToAggSearchQueryRecord.put(recordMovedFromMinToMax.getGroupingId(), new Tuple<>(recordMovedFromMinToMax, false));
    }
}
```